### PR TITLE
WT-11617 Extend WT functional model to support RTS 

### PR DESCRIPTION
**Pull request by pmacko86**
Here is a PR for the model of RTS itself. I hope this PR is not as long as the previous ones...

As usual, feel free to skip or just skim the testing code under `test/model/test`, as it is just a test for a test :)

### CODE REVIEW

File: test/model/src/core/kv_database.cpp [hunk 1]
```diff
@@ -173,4 +173,130 @@ kv_database::txn_snapshot_nolock(txn_id_t do_not_exclude)
       _last_transaction_id, std::move(active_txn_ids));
 }
 
+/*
+ * kv_database::clear --
+ *     Clear the contents of the database.
+ */
+void
+kv_database::clear()
```
Comment (korteland) [inline test/model/src/core/kv_database.cpp:181]: I think I'm missing something - is this called from anywhere?

File: test/model/src/driver/debug_log_parser.cpp [hunk 2]
```diff
@@ -536,10 +582,39 @@ from_debug_log_helper(WT_SESSION_IMPL *session, WT_ITEM *rawrec, WT_LSN *lsnp, W
         break;
     }
 
+    case WT_LOGREC_SYSTEM: {
```
Comment (tod-johnson-mongodb) [inline test/model/src/driver/debug_log_parser.cpp:585]: New line 528 returns 0 for error.  Other similar new lines (538, 549, etc.) return ret.  Why is new line 528 different?

File: test/model/src/driver/debug_log_parser.cpp [hunk 3]
```diff
@@ -644,13 +722,33 @@ debug_log_parser::from_json(kv_database &database, const char *path)
...
         if (log_entry_type == "checkpoint" || log_entry_type == "file_sync" ||
-          log_entry_type == "message" || log_entry_type == "system")
+          log_entry_type == "message")
             continue;
 
         throw model_exception("Unsupported log entry type \"" + log_entry_type + "\"");
     }
+
+    /* Simulate the database starting up. */
+    database.start();
```
Comment (tod-johnson-mongodb) [inline test/model/src/driver/debug_log_parser.cpp:751]: New line 458 and new line 751 both startup the database. Why are both needed? Could both ever be called for the same database? Perhaps comments giving the use case are needed on both of them.

File: test/model/src/include/model/kv_database.h [hunk 4]
```diff
@@ -179,18 +179,97 @@ class kv_database {
...
+
+    /*
+     * kv_database::start_nolock --
+     *     Simulate starting WiredTiger, assuming the locks are held.
+     */
+    void start_nolock();
+
 private:
-    mutable std::mutex _tables_lock;
+    mutable std::recursive_mutex _tables_lock;
```
Comment (tod-johnson-mongodb) [inline test/model/src/include/model/kv_database.h:269]: Why does this lock need to be recursive? Perhaps add a comment.

File: test/model/src/include/model/kv_database.h [hunk 5]
```diff
@@ -179,18 +179,97 @@ class kv_database {
...
     std::unordered_map<std::string, kv_table_ptr> _tables;
 
-    mutable std::mutex _transactions_lock;
+    mutable std::recursive_mutex _transactions_lock;
```
Comment (tod-johnson-mongodb) [inline test/model/src/include/model/kv_database.h:272]: Why does this lock need to be recursive? Perhaps add a comment.

**Review by korteland — COMMENTED** [review]
Mostly LGTM - overall shape of it is good. One question.

File: test/model/test/common/include/model/test/subprocess.h [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+#include <sys/wait.h>
+#include <signal.h>
+#include <string>
+
+extern "C" {
+#include "test_util.h"
+}
+
+/*
+ * in_subprocess_abort --
```
Comment (tod-johnson-mongodb) [inline test/model/test/common/include/model/test/subprocess.h:42]: Title wrong.  Perhaps add to comment " - and normal exit at the end" like in_subprocess_abort.

File: test/model/test/common/util.cpp [hunk 7]
```diff
@@ -63,3 +67,49 @@ create_tmp_file(const char *dir, const char *prefix, const char *suffix)
...
+
+    /* Verify using the debug log. */
+    model::kv_database db_from_debug_log;
+    model::debug_log_parser::from_debug_log(db_from_debug_log, conn);
+    for (auto &t : tables)
+        testutil_assert(db_from_debug_log.table(t.c_str())->verify_noexcept(conn));
+
+    /* Print the debug log to JSON. */
+    std::string tmp_json = create_tmp_file(home, "debug-log-", ".json");
+    wt_print_debug_log(conn, tmp_json.c_str());
```
Comment (tod-johnson-mongodb) [inline test/model/test/common/util.cpp:94]: How do you know that the WiredTiger debug log is the same at both new line 88 and new line 94? This code assumes that is true. Perhaps you should call wt_print_debug_log() all the following places: before new line 88, at new line 94, and after new line 101.  Then compare all the JSON files to make sure they are the same.

File: test/model/src/driver/debug_log_parser.cpp (hunk 2 above)
Comment (pmacko86) [reply test/model/src/driver/debug_log_parser.cpp:585]: Fixed. Nice catch!

File: test/model/src/driver/debug_log_parser.cpp (hunk 3 above)
Comment (pmacko86) [reply test/model/src/driver/debug_log_parser.cpp:751]: I clarified this in the comments, thank you! Line 458 handles database restarts recorded in the log. Line 751 handles the startup that would happen after the model is populated from the debug logs, when the database is open for verification.

File: test/model/test/common/include/model/test/subprocess.h (hunk 6 above)
Comment (pmacko86) [reply test/model/test/common/include/model/test/subprocess.h:42]: Nice catch, thanks.

**Review by tod-johnson-mongodb — APPROVED** [review]
LGTM

File: test/model/src/core/kv_database.cpp (hunk 1 above)
Comment (pmacko86) [reply test/model/src/core/kv_database.cpp:181]: That's right—it's not being used from anywhere, just the _nolock version. I can remove it.

Comment (korteland) [reply test/model/src/core/kv_database.cpp:181]: Yes please - it's not difficult to write if we ever need it.

**Review by korteland — APPROVED** [review]
LGTM!

**Pull request merged**